query-engine/connectors/query-connector/src/filter/scalar.rs

```diff
@@ -15,6 +15,55 @@ pub struct ScalarFilter {
     pub condition: ScalarCondition,
 }
 
+const BATCH_SIZE: usize = 5000;
+
+impl ScalarFilter {
+    pub fn can_batch(&self) -> bool {
+        match self.condition {
+            ScalarCondition::In(ref l) if l.len() > BATCH_SIZE => true,
+            ScalarCondition::NotIn(ref l) if l.len() > BATCH_SIZE => true,
+            _ => false,
+        }
+    }
+
+    pub fn batched(self) -> Vec<ScalarFilter> {
+        fn inner(list: PrismaListValue) -> Vec<PrismaListValue> {
+            let mut batches = Vec::with_capacity(list.len() % BATCH_SIZE + 1);
+            batches.push(Vec::with_capacity(BATCH_SIZE));
+
+            for (idx, item) in list.into_iter().enumerate() {
+                if idx != 0 && idx % BATCH_SIZE == 0 {
+                    batches.push(Vec::with_capacity(BATCH_SIZE));
+                }
+
+                batches.last_mut().unwrap().push(item);
+            }
+
+            batches
+        }
```

Comment:
I was thinking a bit more about this, and if the list here is large (which it is if we are batching), then we are iterating over it, but if we computed the indexes for the batches, we could iterate over those and push the batches one at a time (instead of an element at a time). Here we have per-element costs like the conditional, and the `last_mut` lookup (with bounds check). But that's an optimization we can do later once we have decided we do want to do this this way and the API settles down.

Reply:
Yeah this thing here also is not the final version. It's the first version that somehow works. Now we need some conversation about this. The first limit I'd like to lift is the problem of this not working if we have more than one `IN` statement in the filters.